server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java

```diff
@@ -0,0 +1,343 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.action.search;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.search.TopDocs;
+import org.elasticsearch.action.search.SearchPhaseController.TopDocsStats;
+import org.elasticsearch.common.io.stream.DelayableWriteable;
+import org.elasticsearch.common.io.stream.DelayableWriteable.Serialized;
+import org.elasticsearch.common.io.stream.NamedWriteableRegistry;
+import org.elasticsearch.common.lucene.search.TopDocsAndMaxScore;
+import org.elasticsearch.common.util.concurrent.AbstractRunnable;
+import org.elasticsearch.search.SearchPhaseResult;
+import org.elasticsearch.search.SearchShardTarget;
+import org.elasticsearch.search.aggregations.InternalAggregation.ReduceContextBuilder;
+import org.elasticsearch.search.aggregations.InternalAggregations;
+import org.elasticsearch.search.query.QuerySearchResult;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+
+import static org.elasticsearch.action.search.SearchPhaseController.mergeTopDocs;
+import static org.elasticsearch.action.search.SearchPhaseController.setShardIndex;
+
+/**
+ * A {@link ArraySearchPhaseResults} implementation that incrementally reduces aggregation results
+ * as shard results are consumed.
+ * This implementation can be configured to batch up a certain amount of results and reduce
+ * them asynchronously in the provided {@link Executor} iff the buffer is exhausted.
+ */
+class QueryPhaseResultConsumer extends ArraySearchPhaseResults<SearchPhaseResult> {
+    private static final Logger logger = LogManager.getLogger(SearchPhaseController.class);
+
+    private final Executor executor;
+    private final SearchPhaseController controller;
+    private final SearchProgressListener progressListener;
+    private final ReduceContextBuilder aggReduceContextBuilder;
+    private final NamedWriteableRegistry namedWriteableRegistry;
+
+    private final int topNSize;
+    private final int bufferSize;
+    private final boolean hasTopDocs;
+    private final boolean hasAggs;
+    private final boolean performFinalReduce;
+
+    private final TopDocsStats topDocsStats;
+    private volatile int numReducePhases;
+    private List<SearchShard> processedShards = new ArrayList<>();
+    private boolean hasPartialReduce;
```

Comment:
I think it is worth a comment about why you can't use `reducedAggs != null`. I believe the reason for that is that you want to set this to true before the reduced aggs are ready.

Reply:
++, I pushed https://github.com/elastic/elasticsearch/pull/58461/commits/1fa48801f433588a8a0de2e64851c1e145cf0640